### DISCUSSION TITLE
jotai-scheduler

### DISCUSSION THREAD

**Discussion by L-Qun** [pmndrs/jotai#2710] [Ideas] [answered]
@dai-shi 

Hi Daishi, I created a package called jotai-scheduler: https://github.com/L-Qun/jotai-scheduler

It is used to solve the issue of state update priorities. Jotai helps us address the problem of re-rendering, but there may be different priorities in the rendering order of components. Therefore, I invented jotai-scheduler. It is still in its early stages; please let me know your thoughts.

All components share the same state, but each component has a different level of importance. For instance, the `<Content />` component should be considered more important. Therefore, when this state is updated, the `<Content />` component should be rendered first to provide a better user experience.

Before use `useAtomValue`/`useAtom`, It might appear somewhat laggy:

![20240820225900_rec_](https://github.com/user-attachments/assets/32aec579-2c55-4646-b9ed-3bfb3b07ebd0)

Now we use `useAtomValueWithScheduler`/`useAtomWithScheduler`, the `<Content />` Component is updated first, followed by the `<Sidebar />`, and finally the `<Header />` and `<Footer />`:

![20240820225120_rec_](https://github.com/user-attachments/assets/7e707f81-d4df-47ae-86ac-60da272ec0e8)

Important content will be displayed to users more quickly, thereby providing a better user experience.

The only difference from the native Jotai API is that it requires passing a `priority` parameter, which can be `ImmediatePriority`, `NormalPriority`, or `LowPriority`.

```
const num = useAtomValue(anAtom, {});
```

```
const num = useAtomValueWithSchedule(anAtom, {
  priority: LowPriority,
});
```

**Comment by L-Qun** [comment]
Can this library be added to Jotai Labs https://github.com/jotaijs?

**Reply by dai-shi** [reply]
Sure, I can invite you.

**Comment by dai-shi** [comment]
While I like the idea of scheduler and/or batching for Jotai.
This looks like a bit different from what I expect.
It feels like you are re-inventing something React has already invented (I guess you know it as you name such priorities). Can you try and see if `useDefferedValue` helps for your case? I'm not sure if it works out of the box though.

**Reply by L-Qun** [reply]
Thank you for your reply.

We can use `useDeferredValue` as an alternative, but there are some issues with this approach. 

1. It requires a React version greater than 18. 
2. It does not differentiate between multiple priorities; `useDeferredValue` simply lowers the rendering priority, whereas `useAtomValueWithSchedule` offers a range of rendering priorities.
3. More re-render times: For instance, sometimes we have many items, but we hope the items in the middle of the screen are rendered first, and the code might looks like this:

```
const Item = ({ value }) => {
  const deferredValue = useDeferredValue(value)
  const isOnScreen = ....
  return isOnScreen ? value : deferredValue
}

const Content = () => {
  const itemList = ....

  return (
    <>
      {itemList.map(value => <Item value={value} />)
    </>
  )
}
```

 Some items use `deferredValue`, but others do not, which leads to an increased number of re-renders. Because we used `useDeferredValue`, regardless of whether this value was utilized or not.

@dai-shi please share more of your thoughts with me, thanks!

**Reply by dai-shi** [reply]
1: I don't think it's much of an issue
2: Yeah, but two would simple enough and covers say 95% of use cases. It's more important to use/learn what React provides.
3: I'm not sure if I understand it, but such cases would be good with multiple atoms. I'd use splitAtom for example.

That said, I'm not against having another solution, even if it's for 5% or for React 17.

**Reply by L-Qun** [reply]
For the third point, if all the items use the same atom value and this value changes, it will re-render all the items. However, the items may have different priorities. For instance, the items on the screen might be more important, etc.

**Reply by dai-shi** [reply]
> if all the items use the same atom value and this value changes,

Do all items change at once?

**Reply by L-Qun** [reply]
Yes, If all items are re-rendered, it might cause lag. Batch rendering is a good choice.

**Reply by dai-shi** [reply]
I mean if some items are changed, separating atoms would be better. If all items are changed at once, but some of them are rendered with a low priority, it sounds like a very rare case.

Wait, you are not saying that, because even with your suggestion, it's not possible.

**Reply by dai-shi** [reply]
So, in summary, we should first reach out for `useDefferedValue` or `useTransition`. In some cases, they might not work, and a third-party library could help.

**Reply by L-Qun** [reply, marked as answer]
I agree with you. If users find that `useDeferredValue` or `useTransition` cannot solve their problems, or if they are using React 17, this library can be helpful.
